### ISSUE TITLE
On Length Normalization in the Code

### ISSUE THREAD

**Issue by jc-ryan** [princeton-nlp/SimPO#4]
Congratulations on creating such elegant and effective work! Could you please indicate where the length normalization of the response log probability, as mentioned in the paper, is implemented in the code? I have checked the simplepo_trainer but couldn't find this part of the code. Thank you for your response.

`def simpo_loss(
        self,
        policy_chosen_logps: torch.FloatTensor,
        policy_rejected_logps: torch.FloatTensor,
    ) -> Tuple[torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
        """Compute the SimPO loss for a batch of policy model log probabilities.

        Args:
            policy_chosen_logps: Log probabilities of the policy model for the chosen responses. Shape: (batch_size,)
            policy_rejected_logps: Log probabilities of the policy model for the rejected responses. Shape: (batch_size,)

        Returns:
            A tuple of three tensors: (losses, chosen_rewards, rejected_rewards).
            The losses tensor contains the SimPO loss for each example in the batch.
            The chosen_rewards and rejected_rewards tensors contain the rewards for the chosen and rejected responses, respectively.
        """
        pi_logratios = policy_chosen_logps - policy_rejected_logps
        gamma_logratios = self.gamma / self.beta 
        pi_logratios = pi_logratios.to(self.accelerator.device)
        logits = pi_logratios - gamma_logratios

        if self.loss_type == "sigmoid":
            losses = (
                -F.logsigmoid(self.beta * logits) * (1 - self.label_smoothing)
                - F.logsigmoid(-self.beta * logits) * self.label_smoothing
            )
        elif self.loss_type == "hinge":
            losses = torch.relu(1 - self.beta * logits)
        else:
            raise ValueError(
                f"Unknown loss type: {self.loss_type}. Should be one of ['sigmoid', 'hinge']"
            )

        chosen_rewards = self.beta * policy_chosen_logps.to(self.accelerator.device).detach()
        rejected_rewards = self.beta * policy_rejected_logps.to(self.accelerator.device).detach()

        return losses, chosen_rewards, rejected_rewards`

**Comment by hiyouga** [comment]
Perhaps here: https://github.com/princeton-nlp/SimPO/blob/4e5d6d7ea16828770eb429ef33983fbf2f96b389/scripts/simpo_trainer.py#L89

**Comment by yumeng5** [comment]
> Perhaps here:
> 
> https://github.com/princeton-nlp/SimPO/blob/4e5d6d7ea16828770eb429ef33983fbf2f96b389/scripts/simpo_trainer.py#L89

Yes, this is correct!

**Comment by jc-ryan** [comment]
> > Perhaps here:
> > https://github.com/princeton-nlp/SimPO/blob/4e5d6d7ea16828770eb429ef33983fbf2f96b389/scripts/simpo_trainer.py#L89
> 
> Yes, this is correct!

I got it, thanks.
